src/gtp_context_reg.erl

```diff
@@ -215,14 +219,38 @@ maybe_init_rnd([]) ->
 maybe_init_rnd([{_, RndState}]) ->
     RndState.
 
-alloc_tei(_RndStateKey, _RndState, _KeyFun, 0) ->
+alloc_tei(_RndStateKey, _RndState, _KeyFun, 0, _) ->
     {error, no_tei};
-alloc_tei(RndStateKey, RndState0, KeyFun, Cnt) ->
-    {TEI, RndState} = rand:uniform_s(16#fffffffe, RndState0),
+alloc_tei(RndStateKey, RndState0, KeyFun, Cnt, #{prefix := Prefix0, length := PrefixLen} = Data) ->
+    {TEI0, RndState} = rand:uniform_s(16#fffffffe, RndState0),
+    Mask = 16#ffffffff bsr PrefixLen,
+    Prefix = Prefix0 bsl (?MAX_TRIES - PrefixLen),
+    TEI = Prefix + (TEI0 band Mask),
     case lookup(KeyFun(TEI)) of
 	undefined ->
 	    true = ets:insert(?SERVER, {RndStateKey, RndState}),
 	    {ok, TEI};
 	_ ->
-	    alloc_tei(RndStateKey, RndState, KeyFun, Cnt - 1)
+	    alloc_tei(RndStateKey, RndState, KeyFun, Cnt - 1, Data)
     end.
+
+get_teid_cfg() ->
+	get_teid_pref_len(application:get_env(ergw, teid), #{prefix => 0, length => 1}).
+
+get_teid_pref_len({ok, [_|_] = TeidCfg}, DefTEID) ->
+	case proplists:get_value(length, TeidCfg, 0) of
```

Comment:
It is easier to read code that does not nest case. Perhaps split this into two help functions?